packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt

```diff
@@ -512,15 +510,136 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
 
 
     @VisibleForTesting
-    fun deserializeNestedModels(serializedModelData: Map<String, Any>): Map<String, Any> {
-        return serializedModelData.mapValues {
-            if (it.value is Map<*, *>) {
-                SerializedModel.builder()
-                    .serializedData(deserializeNestedModels(it.value as HashMap<String, Any>))
-                    .modelSchema(null)
+    fun deserializeNestedModels(serializedModelData: Map<String, Any?>, modelSchema: ModelSchema): Map<String, Any?> {
+        val result = mutableMapOf<String, Any?>()
+
+        // iterate over schema fields to deserialize each field value
+        for ((key, field) in modelSchema.fields.entries) {
+            // ignore if serializedModelData doesn't contain value for the current field key
+            if (!serializedModelData.containsKey(key)) {
+                continue
+            }
+            val fieldSerializedData = serializedModelData[key]
+
+            // if the filed serialized value is null
+            // assign null to this field in the deserialized map
+            if (fieldSerializedData == null) {
+                result[key] = fieldSerializedData
+                continue
+            }
+
+            if (field.isModel) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldModelSchema = modelProvider.modelSchemas()[field.targetType] ?: continue
+                result[key] = SerializedModel.builder()
+                    .serializedData(deserializeNestedModels(fieldSerializedData as Map<String, Any?>, fieldModelSchema))
+                    .modelSchema(fieldModelSchema)
                     .build()
-            } else
-                it.value
+            } else if (field.isCustomType) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldCustomTypeSchema = modelProvider.customTypeSchemas()[field.targetType] ?: continue
+
+                // When a field is custom type
+                // the field value can only be a single custom type
+                // or a list of item of the same custom type
+                if (field.isArray) {
+                    result[key] = (fieldSerializedData as List<Map<String, Any?>>).map { listItem ->
+                        SerializedCustomType.builder()
+                            .serializedData(deserializeNestedCustomType(listItem, fieldCustomTypeSchema))
+                            .customTypeSchema(fieldCustomTypeSchema)
+                            .build()
+                    }
+                } else {
+                    result[key] = SerializedCustomType.builder()
+                        .serializedData(
+                            deserializeNestedCustomType(fieldSerializedData as Map<String, Any>, fieldCustomTypeSchema)
+                        )
+                        .customTypeSchema(fieldCustomTypeSchema)
+                        .build()
```

Comment:
The code seems duplicate but its legit just invoking class builder. It's just like invoke `new SerializeCustomType()` twice. Extract this may not worth creating a new private method...